trytond/doc/topics/configuration.rst
.. _topics-configuration:

==================
Configuration file
==================

The configuration file controls some aspects of the behavior of Tryton.
The file uses a simple ini-file format. It consists of sections, led by a
``[section]`` header and followed by ``name = value`` entries:

.. highlight:: ini

::

    [database]
    uri = postgresql://user:password@localhost/
    path = /var/lib/trytond

For more information see ConfigParser_.

.. _ConfigParser: http://docs.python.org/2/library/configparser.html

The default value of any option can be changed using environment variables
with names using this syntax: ``TRYTOND_<SECTION>__<NAME>``.

Sections
========

This section describes the different main sections that may appear in a Tryton
configuration file, the purpose of each section, its possible keys, and their
possible values.
Some modules could request the usage of other sections for which the guideline
asks them to be named like their module.

.. contents::
   :local:
   :backlinks: entry
   :depth: 2

.. _config-web:

web
---

Defines the behavior of the web interface.

.. _config-web.listen:

listen
~~~~~~

Defines the couple of host (or IP address) and port number separated by a colon
to listen on.

Default ``localhost:8000``

.. note::
   To listen on all IPv4 interfaces use the value ``0.0.0.0:8000`` and for all
   IPv6 interfaces use ``[::]:8000``.

.. _config-web.hostname:

hostname
~~~~~~~~

Defines the hostname to use when generating a URL when there is no request
context available, for example during a cron job.

.. _config-web.root:

root
~~~~

Defines the root path served by ``GET`` requests.

Default: Under the ``www`` directory of user's home running ``trytond``.

.. _config-web.num_proxies:

num_proxies
~~~~~~~~~~~

The number of proxy servers in front of ``trytond``.

Default: 0

.. _config-web.cache_timeout:

cache_timeout
~~~~~~~~~~~~~

The cache timeout in seconds.

Default: 12h

.. _config-web.cors:

cors
~~~~

The list (one per line) of origins allowed for `Cross-Origin Resource sharing
<https://en.wikipedia.org/wiki/Cross-origin_resource_sharing>`_.
For example::

   cors =
      http://example.com
      https://example.com

.. _config-web.avatar_base:

avatar_base
~~~~~~~~~~~

The base URL without a path for avatar URL.

Default: ``''``

.. note:: It can be used to setup a CDN.


.. _config-web.avatar_timeout:

avatar_timeout
~~~~~~~~~~~~~~

The time in seconds that the avatar can be stored in cache.

Default: 7 days

.. _config-database:

database
--------

Defines how the database is managed.

.. _config-database.uri:

uri
~~~

Contains the URI to connect to the SQL database. The URI follows the :rfc:`3986`.
The typical form is:

    database://username:password@host:port/?param1=value1&param2=value2

The parameters are database dependent, check the database documentation for a
list of valid parameters.

Default: The value of the environment variable ``TRYTOND_DATABASE_URI`` or
``sqlite://`` if not set.

The available databases are:

PostgreSQL
**********

``psycopg2`` supports two type of connections:

- TCP/IP connection: ``postgresql://user:password@localhost:5432/``
- Unix domain connection:
   - with password authentication: ``postgresql://username:password``
   - with operating system user name: ``postgresql://``

Please refer to `psycopg2 for the complete specification of the URI
<https://www.psycopg.org/docs/module.html#psycopg2.connect>`_.

A list of parameters supported by PostgreSQL can be found in the
`documentation
<https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-PARAMKEYWORDS>`__.

.. note::
   ``fallback_application_name`` parameter from aforementioned documentation can
   be set directly thanks to the ``TRYTOND_APPNAME`` environment variable.

SQLite
******

The URI is defined as ``sqlite://``

If the name of the database is ``:memory:``, the parameter ``mode`` will be set
to ``memory`` thus using a pure in-memory database.

The recognized query parameters can be found in SQLite's
`documentation
<https://www.sqlite.org/uri.html#recognized_query_parameters>`__.

.. _config-database.path:

path
~~~~

The directory where Tryton stores files and so the user running
:command:`trytond` must have write access on this directory.

Default: The :file:`db` folder under the user home directory running
:command:`trytond`.

.. _config-database.list:

list
~~~~

A boolean value to list available databases.

Default: ``True``

.. _config-database.retry:

retry
~~~~~

The number of retries when a database operational error occurs during a request.

Default: ``5``

.. _config-database.subquery_threshold:

subquery_threshold
~~~~~~~~~~~~~~~~~~

The number of records in the target relation under which a sub-query is used.

Default: ``1000``

.. _config-database.language:

language
~~~~~~~~

The main language of the database that will be used for storage in the main
table for translations.

Default: ``en``

.. _config-database.avatar_filestore:

avatar_filestore
~~~~~~~~~~~~~~~~

This configuration value indicates whether the avatars should be stored in the
:py:mod:`trytond.filestore` (``True``) or the database (``False``).

Default: ``False``

.. _config-database.avatar_prefix:

avatar_prefix
~~~~~~~~~~~~~

The prefix to use with the :ref:`FileStore <ref-filestore>` to store avatars.

Default: ``None``

.. _config-database.default_name:

default_name
~~~~~~~~~~~~

The name of the database to use for operations without a database name.
Default: ``template1`` for PostgreSQL, ``:memory:`` for SQLite.

.. _config-database.timeout:

timeout
~~~~~~~

The timeout duration in seconds after which the connections to unused databases
are closed.
Default: ``1800`` (30 minutes)

.. _config-database.minconn:

minconn
~~~~~~~

The minimum number of connections to keep in the pool (if the backend supports
pool) per process.
Default: ``1``

.. _config-database.maxconn:

maxconn
~~~~~~~

The maximum number of simultaneous connections to the database per process.
Default: ``64``

.. _config-database.unaccent_function:

unaccent_function
~~~~~~~~~~~~~~~~~

The name of the unaccent function.

Default: ``unaccent``

.. _config-database.similarity_function:

similarity_function
~~~~~~~~~~~~~~~~~~~

The name of the similarity function.

Default: ``similarity``

binary_scanner
~~~~~~~~~~~~~~

The command used to scan the content of Binary fields.
The command receive the quarantine directory as last argument.
If the command returns a non zero status, it is considered that at least one
file in the directory is malicious.

binary_scanner_directory
~~~~~~~~~~~~~~~~~~~~~~~~

The directory where the binary content are stored before being checked
by the scanner.

Default: The temporary directory as determined by Python's tempfile_ module.

.. _tempfile: https://docs.python.org/3/library/tempfile.html#tempfile.gettempdir

.. _config-request:

request
-------

.. _config-request.max_size:

max_size
~~~~~~~~

The maximum size in bytes of unauthenticated request (zero means no limit).

Default: 2MB

.. _config-request.max_size_authenticated:

max_size_authenticated
~~~~~~~~~~~~~~~~~~~~~~

The maximum size in bytes of an authenticated request (zero means no limit).

Default: 2GB

.. _config-request.timeout:

timeout
~~~~~~~

The timeout in seconds before aborting requests that have their execution time
depending on the parameters.

Default: ``60``

.. _config-request.records_limit:

records_limit
~~~~~~~~~~~~~

The maximal number of records processed by requests.

Default: ``None``

.. _config-cache:

cache
-----

Defines size of various cache.

.. _config-cache.transaction:

transaction
~~~~~~~~~~~

The number of contextual caches kept per transaction.

Default: ``10``

.. _config-cache.model:

model
~~~~~

The number of different model kept in the cache per transaction.

Default: ``200``

.. _config-cache.record:

record
~~~~~~

The number of record loaded kept in the cache of the list.
It can be changed locally using the ``_record_cache_size`` key in
:attr:`Transaction.context <trytond.transaction.Transaction.context>`.

Default: ``2000``

.. _config-cache.field:

field
~~~~~

The number of field to load with an ``eager`` :attr:`Field.loading
<trytond.model.fields.Field.loading>`.

Default: ``100``

.. _config-cache.default:

default
~~~~~~~

The default :attr:`~trytond.cache.Cache.size_limit` of :class:`~trytond.cache.Cache`.

Default: ``1024``

.. _config-cache.clean_timeout:

clean_timeout
~~~~~~~~~~~~~

The minimum number of seconds between two cleanings of the cache.
If the value is 0, the notification between processes will be done using
channels if the back-end supports them.

Default: ``300``

.. _config-cache.select_timeout:

select_timeout
~~~~~~~~~~~~~~

The timeout duration of the select call when using channels.

Default: ``60``

.. _config-cache.count_timeout:

count_timeout
~~~~~~~~~~~~~

The cache timeout duration in seconds of the estimation of records.

Default: ``86400`` (1 day)

.. _config-cache.count_clear:

count_clear
~~~~~~~~~~~

The number of operations after which the counting estimation of records is
cleared.

Default: ``1000``

.. _config-cron:

cron
----

.. _config-cron.clean_days:

clean_days
~~~~~~~~~~

The number of days after which scheduled task logs are removed.

Default: ``30``

.. _config-queue:

queue
-----

.. _config-queue.worker:

worker
~~~~~~

Activate asynchronous processing of the tasks. Otherwise they are performed at
the end of the requests.

Default: ``False``

.. _config-queue.clean_days:

clean_days
~~~~~~~~~~

The number of days after which processed tasks are removed.

Default: ``30``

.. _config-queue.batch_size:

batch_size
~~~~~~~~~~

The default number of the instances to process in a batch.

Default: ``20``

.. _config-error:

error
-----

.. _config-error.clean_days:

clean_days
~~~~~~~~~~

The number of days after which reported errors are removed.

Default: ``90``

.. _config-table:

table
-----

This section allows to override the default generated table name for a
:class:`~trytond.model.ModelSQL`.
The main goal is to bypass limitation on the name length of the database
backend.
For example::

    [table]
    account.invoice.line = acc_inv_line
    account.invoice.tax = acc_inv_tax

.. _config-ssl:

ssl
---

Activates SSL_ on the web interface.

.. note:: It is recommended to delegate the SSL support to a proxy.

.. _config-ssl.privatekey:

privatekey
~~~~~~~~~~

The path to the private key.

.. _config-ssl.certificate:

certificate
~~~~~~~~~~~

The path to the certificate.

.. tip::
   Set only one of ``privatekey`` or ``certificate`` to ``true`` if the SSL is
   delegated.

.. _config-email:

email
-----

.. note:: Email settings can be tested with the ``trytond-admin`` command

.. _config-email.uri:

uri
~~~

The SMTP-URL_ to connect to the SMTP server which is extended to support SSL_
and STARTTLS_.
The available protocols are:

    - ``smtp``: simple SMTP
    - ``smtp+tls``: SMTP with STARTTLS
    - ``smtps``: SMTP with SSL

The uri accepts the following additional parameters:

* ``local_hostname``: used as FQDN of the local host in the HELO/EHLO commands,
  if omited it will use the value of ``socket.getfqdn()``.
* ``timeout``: A number of seconds used as timeout for blocking operations. A
  ``socket.timeout`` will be raised when exceeded. If omited the default timeout
  will be used.


Default: ``smtp://localhost:25``

.. _config-email.from:

from
~~~~

Defines the default ``From`` address (using :rfc:`5322`) for emails sent by
Tryton.

For example::

    from: "Company Inc" <info@example.com>

Default: The login name of the :abbr:`OS (Operating System)` user.

.. _config-email.retry:

retry
~~~~~

The number of retries when the SMTP server returns a temporary error.

Default: ``5``

.. _config-session:

session
-------

.. _config-session.authentications:

authentications
~~~~~~~~~~~~~~~

A comma separated list of the authentication methods to try when attempting to
verify a user's identity. Each method is tried in turn, following the order of
the list, until one succeeds. In order to allow `multi-factor authentication`_,
individual methods can be combined together using a plus (``+``) symbol.

Example::

    authentications = password+sms,ldap

Each combined method can have options to skip them if they are met except for
the first method.
They are defined by appending their name to the method name after a question
mark (``?``) and separated by colons (``:``).

Example::

   authentications = password+sms?ip_address:device_cookie


By default, Tryton only supports the ``password`` method.  This method compares
the password entered by the user against a stored hash of the user's password.
By default, Tryton supports the ``ip_address`` and ``device_cookie`` options.
The ``ip_address`` compares the client IP address with the known network list
defined in `authentication_ip_network`_.
The ``device_cookie`` checks the client device is a known device of the user.
Other modules can define additional authentication methods and options, please
refer to their documentation for more information.

Default: ``password``

.. _config-session.authentication_ip_network:

authentication_ip_network
~~~~~~~~~~~~~~~~~~~~~~~~~

A comma separated list of known IP networks used to check for ``ip_address``
authentication method option.

Default: ``''``

.. _config-session.max_age:

max_age
~~~~~~~

The time in seconds that a session stay valid.

Default: ``2592000`` (30 days)

.. _config-session.timeout:

timeout
~~~~~~~

The time in seconds without activity before the session is no more fresh.

Default: ``300`` (5 minutes)

.. _config-session.max_attempt:

max_attempt
~~~~~~~~~~~

The maximum authentication attempt before the server answers unconditionally
``Too Many Requests`` for any other attempts. The counting is done on all
attempts over a period of ``timeout``.

Default: ``5``

.. _config-session.max_attempt_ip_network:

max_attempt_ip_network
~~~~~~~~~~~~~~~~~~~~~~

The maximum authentication attempt from the same network before the server
answers unconditionally ``Too Many Requests`` for any other attempts. The
counting is done on all attempts over a period of ``timeout``.

Default: ``300``

.. _config-session.ip_network_4:

ip_network_4
~~~~~~~~~~~~

The network prefix to apply on IPv4 address for counting the authentication
attempts.

Default: ``32``

.. _config-session.ip_network_6:

ip_network_6
~~~~~~~~~~~~

The network prefix to apply on IPv6 address for counting the authentication
attempts.

Default: ``56``

.. _config-password:

password
--------

.. _config-password.length:

length
~~~~~~

The minimal length required for the user password.

Default: ``8``

.. _config-password.forbidden:

forbidden
~~~~~~~~~

The path to a file containing one forbidden password per line.

.. _config-password.reset_timeout:

reset_timeout
~~~~~~~~~~~~~

The time in seconds until the reset password expires.

Default: ``86400`` (24h)

.. _config-attachment:

attachment
----------

Defines how to store the attachments

.. _config-attachment.filestore:

filestore
~~~~~~~~~

A boolean value to store attachment in the :ref:`FileStore <ref-filestore>`.

Default: ``True``

.. _config-attachment.store_prefix:

store_prefix
~~~~~~~~~~~~

The prefix to use with the ``FileStore``.

Default: ``None``

.. _config-bus:

bus
---

.. _config-bus.allow_subscribe:

allow_subscribe
~~~~~~~~~~~~~~~

A boolean value to allow clients to subscribe to bus channels.

Default: ``False``

.. _config-bus.url_host:

url_host
~~~~~~~~

If set redirects bus requests to the host URL.

.. _config-bus.long_polling_timeout:

long_polling_timeout
~~~~~~~~~~~~~~~~~~~~

The time in seconds to keep the connection to the client opened when using long
polling for bus messages

Default: ``300``

.. _config-bus.cache_timeout:

cache_timeout
~~~~~~~~~~~~~

The number of seconds a message should be kept by the queue before being
discarded.

Default: ``300``

.. _config-bus.select_timeout:

select_timeout
~~~~~~~~~~~~~~

The timeout duration of the select call when listening on a channel.

Default: ``5``

.. _config-report:

report
------

.. _config-report.convert_command:

convert_command
---------------

The command to convert document between formats.

The available keywords are:

   - ``%(directory)s``: the temporary working directory
   - ``%(input_format)s``: the format of the file to convert
   - ``%(input_extension)s``: the extension of the file to convert
   - ``%(input_path)s``: the path of the file to convert
   - ``%(output_format)s``: the format to which the file must be converted
   - ``%(output_extension)s``: the extension for the converted file
   - ``%(output_path)s``: the path where the converted file must be written

The command must write the result in ``%(output_path)s``.

.. _config-html:

html
----

.. _config-html.src:

src
~~~

The URL pointing to `TinyMCE <https://www.tiny.cloud/>`_ editor.

Default: ``https://cloud.tinymce.com/stable/tinymce.min.js``

.. _config-html.license_key:

license_key
~~~~~~~~~~~

The license key for TinyMCE.

Default: ``gpl``

.. _config-html.plugins:

plugins
~~~~~~~

The space separated list of TinyMCE plugins to load.
It can be overridden for specific models and fields using the names:
``plugins-<model>-<field>`` or ``plugins-<model>``.

Default: ``''``

.. _config-html.css:

css
~~~

The JSON list of CSS files to load.
It can be overridden for specific models and fields using the names:
``css-<model>-<field>`` or ``css-<model>``.

Default: ``[]``

.. _config-html.class:

class
~~~~~

The class to add on the body.
It can be overridden for specific models and fields using the names:
``class-<model>-<field>`` or ``class-<model>``.

Default: ``''``

.. _config-wsgi_middleware:

wsgi middleware
---------------

The section lists the `WSGI middleware`_ class to load.
Each middleware can be configured with a section named ``wsgi <middleware>``
containing ``args`` and ``kwargs`` options.

Example::

    [wsgi middleware]
    ie = werkzeug.contrib.fixers.InternetExplorerFix

    [wsgi ie]
    kwargs={'fix_attach': False}

.. note::
   The options can be set using environment variables with names like:
   ``TRYTOND_WSGI_<MIDDLEWARE>__<NAME>``.


.. _JSON-RPC: http://en.wikipedia.org/wiki/JSON-RPC
.. _XML-RPC: http://en.wikipedia.org/wiki/XML-RPC
.. _SMTP-URL: https://datatracker.ietf.org/doc/html/draft-earhart-url-smtp-00
.. _SSL: http://en.wikipedia.org/wiki/Secure_Sockets_Layer
.. _STARTTLS: http://en.wikipedia.org/wiki/STARTTLS
.. _WSGI middleware: https://en.wikipedia.org/wiki/Web_Server_Gateway_Interface#Specification_overview
.. _`multi-factor authentication`: https://en.wikipedia.org/wiki/Multi-factor_authentication


sao/CHANGELOG
* Use sandboxed iframe to display document (issue14290)
* Display the number of selected records
* Humanize the count result
* Add limit to search_count

Version 6.0.17 - 2022-05-06
* Bug fixes (see mercurial logs for details)

Version 6.0.16 - 2022-04-15
* Bug fixes (see mercurial logs for details)

Version 6.0.15 - 2022-03-01
* Bug fixes (see mercurial logs for details)

Version 6.0.14 - 2022-02-16
* Bug fixes (see mercurial logs for details)

Version 6.0.13 - 2022-02-01
* Bug fixes (see mercurial logs for details)

Version 6.0.12 - 2022-01-15
* Bug fixes (see mercurial logs for details)

Version 6.0.11 - 2022-01-01
* Bug fixes (see mercurial logs for details)

Version 6.0.10 - 2021-12-03
* Bug fixes (see mercurial logs for details)

Version 6.0.9 - 2021-11-04
* Bug fixes (see mercurial logs for details)

Version 6.0.8 - 2021-10-15
* Bug fixes (see mercurial logs for details)

Version 6.0.7 - 2021-10-01
* Bug fixes (see mercurial logs for details)

Version 6.0.6 - 2021-09-16
* Bug fixes (see mercurial logs for details)

Version 6.0.5 - 2021-09-01
* Bug fixes (see mercurial logs for details)

Version 6.0.4 - 2021-08-05
* Bug fixes (see mercurial logs for details)

Version 6.0.3 - 2021-07-21
* Bug fixes (see mercurial logs for details)

Version 6.0.2 - 2021-07-01
* Bug fixes (see mercurial logs for details)

Version 6.0.1 - 2021-06-17
* Bug fixes (see mercurial logs for details)

Version 6.0.0 - 2021-05-03
* Bug fixes (see mercurial logs for details)
* Use appropriate types for email and url widgets
* Add completion on xxx2Many and Dict
* Add attachment preview
* Support document widget
* Support icon of type URL
* Allow sending email with existing attachments
* Add paths and model context to action data
* Add indentation in CSV export of tree
* Manage which records to use for actions
* Add drag and drop on CSV export and import
* Remove tempusdominus dependency
* Handle device cookie
* Add breadcrumb as title of window form
* Manage help for each selection
* Display revision on dialog
* Support empty value for timedelta converter
* Unify PYSON string format

Version 5.8.0 - 2020-11-02
* Bug fixes (see mercurial logs for details)
* Format timedelta in CSV export
* Manage symbol widget
* Fallback to model name as title
* Set treeview footer sticky on bottom
* Add get, in_ and contains methods to PYSON
* Manage deletable and writable state
* Support e-mail template
* Always encode CSV export in UTF-8
* Use tempusdominus and Popper for date time picker
* Support PYSON comparison of date and datetime
* Customize bootstrap default style
* Set field's name as input's name attribute
* Position copied records based on order
* Sanitize translated RichtText fields content (issue9453)
* Allow configuration of default colors for graph and calendar
* Sanitize RichtText fields content (issue9405)
* Escape external string (issue9394)
* Keep context in sessionStorage
* Use existing context for get_preferences
* Escape external strings (issue9351)
* Add context to export URL
* Allow changes to editable tree checkboxes without first selecting the line

Version 5.6.0 - 2020-05-04
* Bug fixes (see mercurial logs for details)
* Support link button on form
* Add URL for the current export
* Add option to export listed records
* Set treeview header sticky on top
* Add noreferrer noopener to external URL (issue9089)
* Manage readonly dates in calendar view
* Manage editable on calendar view
* Manage named separator as label
* Add MultiSelection entry to Dict field
* Position new record based on order
* Reset display_size of TreeView when group is cleared or changed
* Allow relation field to be selected in CSV export
* Hide empty cell on small screen
* Support validate attribute of wizard's Button
* Add infinite scroll

Version 5.4.0 - 2019-11-04
* Bug fixes (see mercurial logs for details)
* Support visual context on tree view
* Support multiselection field
* Support dot notation on PYSON Eval
* Manage sequence
* Add drag and drop on tree views
* Support expand on column
* Set minimal width for columns
* Order Dict keys by sequence
* Support relation fields in domain parser
* Add view modified
* Add Dict widget for tree view

Version 5.2.0 - 2019-05-06
* Bug fixes (see mercurial logs for details)
* Add domain inversion for reference fields
* Add list-form view
* Support HTML widget
* Use locale format for data export
* Add day view on calendar
* Add CSV exports into print toolbar
* Display unread and total of notes
* Use resources method
* Rename roundup.url configuration to bug.url
* Use separate XML parser in Views
* Add range selection with Shift-click
* Use between numbers in filter box
* Fill 'to' with 'from' value on filter box
* Add tab domain in URL
* Use properties
* Allow sao classes to use javascript properties
* Improve layout of CSV import/export
* Add support for height and width on notebook
* Use tab key for tab navigation
* Use record and field properties of widget instead of method arguments
* Reduce offset when no search result
* Add support for plugins
* Add support for expand, height and width attributes
* Add drag & drop support on attachment button

Version 5.0.0 - 2018-10-01
* Bug fixes (see mercurial logs for details)
* Prevent similar tab
* Remove dialog for input file
* Use dropdown for attachment
* Manage icons in input
* Support RTL language direction
* Apply factor on domain parser
* Use tab name in CSV Export/Import
* Add bus management
* Load custom CSS and Javascript
* New icons
* Support timestamp field
* Validate dictionary items
* Add (i)like support to domain inversion
* Improve loading strategy by using views
* Add support of width attribute on tree view
* Use non conflicting keyboard shortcuts
* Use the context to get the suffix of window name
* Call on_change methods after setting default dtstart on calendar view
* Allow to open in new tab Many2One and Reference with CTRL key
* Store session in browser storage
* Increase the space available for tab list

tryton/tryton/config.py
# This file is part of Tryton.  The COPYRIGHT file at the top level of
# this repository contains the full copyright notices and license terms.
import configparser
import datetime
import gettext
import locale
import logging
import optparse
import os
import glob
import pathlib
import shutil
import sys
from tempfile import NamedTemporaryFile

from gi.repository import GdkPixbuf

from tryton import __version__

logger = logging.getLogger(__name__)
_ = gettext.gettext


def _reverse_series_iterator(starting_version):
    major, minor = map(int, starting_version.split('.'))
    while major >= 0 and minor >= 0:
        yield f"{major}.{minor}"
        if minor == 0:
            major -= 1
            minor = 8
        else:
            minor -= 2 if not minor % 2 else 1


def copy_previous_configuration(config_element):
    current_version = __version__.rsplit('.', 1)[0]
    config_dir = get_config_root()
    for version in _reverse_series_iterator(current_version):
        config_path = os.path.join(config_dir, version, config_element)
        if version == current_version and os.path.exists(config_path):
            break
        elif os.path.exists(config_path):
            if os.path.isfile(config_path):
                shutil.copy(config_path, get_config_dir())
            elif os.path.isdir(config_path):
                shutil.copytree(
                    config_path,
                    os.path.join(get_config_dir(), config_element))
            break


def get_config_root():
    if os.name == 'nt':
        appdata = os.environ['APPDATA']
        if not isinstance(appdata, str):
            appdata = str(appdata, sys.getfilesystemencoding())
        config_path = os.path.join(appdata, '.config')
    else:
        config_path = os.path.expanduser(os.getenv(
            'XDG_CONFIG_HOME', os.path.join('~', '.config')))
    return os.path.join(config_path, 'tryton')


def get_config_dir():
    return os.path.join(get_config_root(), __version__.rsplit('.', 1)[0])


if not os.path.isdir(get_config_dir()):
    os.makedirs(get_config_dir(), mode=0o700)
    copy_previous_configuration('tryton.conf')
    copy_previous_configuration('known_hosts')
    copy_previous_configuration('accel_map')
    copy_previous_configuration('profiles.cfg')
    copy_previous_configuration('plugins')


class ConfigManager(object):
    "Config manager"

    def __init__(self):
        demo_server = ''
        demo_database = ''
        self.defaults = {
            'login.profile': demo_server,
            'login.login': 'demo',
            'login.service': '',
            'login.service.port': 8001,
            'login.host': demo_server,
            'login.db': demo_database,
            'login.expanded': False,
            'tip.autostart': False,
            'tip.position': 0,
            'form.toolbar': True,
            'client.title': 'Coog',
            'rpc.cache_size': 1024,
            'client.modepda': False,
            'client.toolbar': 'default',
            'client.save_tree_width': True,
            'client.save_tree_state': False,
            'client.spellcheck': False,
            'client.code_scanner_sound': True,
            'client.lang': locale.getdefaultlocale()[0],
            'client.language_direction': 'ltr',
            'client.email': '',
            # JCA: Set default limit to 100 for performances
            'client.limit': 100,
            'client.check_version': False,
            'client.bus_timeout': 10 * 60,
            'icon.colors': '#0094d2,#57a639,#cc0000',
            'tree.colors': '#777,#dff0d8,#fcf8e3,#f2dede',
            'calendar.colors': '#fff,#3465a4',
            'graph.color': '#3465a4',
            'image.max_size': 10 ** 6,
            'image.cache_size': 1024,
            'bug.url': 'https://support.coopengo.com/',
            'download.frequency': 60 * 60 * 8,
            'menu.pane': 320,
        }
        self.config = {}
        self.options = {}
        self.arguments = []

    def parse(self):
        parser = optparse.OptionParser(version=("Coog %s" % __version__),
                usage="Usage: %prog [options] [url]")
        parser.add_option("-c", "--config", dest="config",
                help=_("specify alternate config file"))
        parser.add_option("-d", "--dev", action="store_true",
                default=False, dest="dev",
                help=_("development mode"))
        parser.add_option("-v", "--verbose", action="store_true",
                default=False, dest="verbose",
                help=_("logging everything at INFO level"))
        parser.add_option("-l", "--log-level", dest="log_level",
                help=_("specify the log level: "
                "DEBUG, INFO, WARNING, ERROR, CRITICAL"))
        parser.add_option("-o", "--log-ouput", dest="log_output", default=None,
            help=_("specify the file used to output logging information"))
        parser.add_option("-u", "--user", dest="login",
                help=_("specify the login user"))
        parser.add_option("-s", "--server", dest="host",
                help=_("specify the server hostname:port"))
        parser.add_option(
            '--no-thread', default=True, action='store_false', dest='thread',
            help=_("disable thread usage"))
        opt, self.arguments = parser.parse_args()
        self.rcfile = opt.config or os.path.join(
            get_config_dir(), 'tryton.conf')
        self.load()

        logging_config = {
            'format': '%(asctime)s.%(msecs)03d:%(levelname)s:%(message)s',
            'datefmt': '%Y-%m-%d %H:%M:%S',
            }
        if opt.log_output:
            logging_config['filename'] = opt.log_output
        else:
            config_dir = pathlib.Path(get_config_dir())
            # Remove old log files, keeping only 10
            files = [
                str(x.absolute())
                for x in config_dir.glob('tryton-*.log')
                if x.is_file()]
            files.sort(reverse=True)
            for file in files[9:]:
                os.remove(file)
            now = datetime.datetime.now().strftime('%Y%m%d-%Hh%Mm%Ss')
            logging_file = config_dir / f'tryton-{now}.log'
            logging_config['filename'] = str(logging_file)

        loglevels = {
            'DEBUG': logging.DEBUG,
            'INFO': logging.INFO,
            'WARNING': logging.WARNING,
            'ERROR': logging.ERROR,
            'CRITICAL': logging.CRITICAL,
            }
        if not opt.log_level:
            if opt.verbose:
                opt.log_level = 'INFO'
            else:
                opt.log_level = 'ERROR'
        logging_config['level'] = loglevels[opt.log_level.upper()]
        logging.basicConfig(**logging_config)

        self.options['dev'] = opt.dev
        for arg in ['login', 'host']:
            if getattr(opt, arg):
                self.options['login.' + arg] = getattr(opt, arg)
        self.options['thread'] = opt.thread

    def save(self):
        try:
            parser = configparser.ConfigParser()
            for entry in list(self.config.keys()):
                if not len(entry.split('.')) == 2:
                    continue
                section, name = entry.split('.')
                if not parser.has_section(section):
                    parser.add_section(section)
                parser.set(section, name, str(self.config[entry]))
            with open(self.rcfile, 'w') as fp:
                parser.write(fp)
        except IOError:
            logger.warn("Unable to write config file %s", self.rcfile)
            return False
        return True

    def load(self):
        parser = configparser.ConfigParser()
        try:
            parser.read([self.rcfile])
        except configparser.Error:
            config_dir = os.path.dirname(self.rcfile)
            with NamedTemporaryFile(
                    delete=False, prefix='tryton_', suffix='.conf',
                    dir=config_dir) as temp_file:
                temp_name = temp_file.name
            shutil.copy(self.rcfile, temp_name)
            logger.error(
                f"Failed to parse {self.rcfile}. "
                f"A backup can be found at {temp_name}", exc_info=True)
            return
        for section in parser.sections():
            for (name, value) in parser.items(section):
                if value.lower() == 'true':
                    value = True
                elif value.lower() == 'false':
                    value = False
                if section == 'client' and name == 'limit':
                    # First convert to float to be backward compatible with old
                    # configuration
                    value = int(float(value))
                self.config[section + '.' + name] = value
        return True

    def __setitem__(self, key, value, config=True):
        self.options[key] = value
        if config:
            self.config[key] = value

    def __getitem__(self, key):
        return self.options.get(key, self.config.get(key,
            self.defaults.get(key)))


CONFIG = ConfigManager()
CURRENT_DIR = os.path.dirname(__file__)
if not isinstance(CURRENT_DIR, str):
    CURRENT_DIR = str(CURRENT_DIR, sys.getfilesystemencoding())

PIXMAPS_DIR = os.path.join(CURRENT_DIR, 'data', 'pixmaps', 'tryton')
if not os.path.isdir(PIXMAPS_DIR):
    try:
        import importlib.resources
        ref = importlib.resources.files('tryton') / 'data/pixmaps/tryton'
        with importlib.resources.as_file(ref) as path:
            PIXMAPS_DIR = path
    except ImportError:
        import pkg_resources
        PIXMAPS_DIR = pkg_resources.resource_filename(
            'tryton', 'data/pixmaps/tryton')
SOUNDS_DIR = os.path.join(CURRENT_DIR, 'data', 'sounds')
if not os.path.isdir(SOUNDS_DIR):
    try:
        import importlib.resources
        ref = importlib.resources.files('tryton') / 'data/sounds'
        with importlib.resources.as_file(ref) as path:
            SOUNDS_DIR = path
    except ImportError:
        import pkg_resources
        SOUNDS_DIR = pkg_resources.resource_filename('tryton', 'data/sounds')

TRYTON_ICON = GdkPixbuf.Pixbuf.new_from_file(
    os.path.join(PIXMAPS_DIR, 'coog_no_text.svg'))
